lib/app_manifest/manifest_validator.rb

```diff
@@ -0,0 +1,142 @@
+module AppManifest
+  class InvalidManifest < StandardError; end
+
+  # app.json manifest validation
+  module ManifestValidator
+
+    def self.validate!(manifest)
+      do_validate(manifest)
+    end
+
+    private
+
+    def self.do_validate(hash, env_name: nil)
+      raise InvalidManifest, "Expected object but got: #{hash.class}" unless hash.is_a?(Hash)
+      hash.each do |key, value|
+        case key.to_s
+        when 'addons'
+          validate_array(key, value, required: false, env_name: env_name) do |v, index: nil, env_name: nil|
+            validate_addon_value(key, v, index: index, env_name: env_name)
+          end
+        when 'buildpacks'
+          validate_array(key, value, required: false, env_name: env_name)
+        when 'env'
+          validate_env(key, value, env_name: env_name)
+        when 'name'
+          validate_string(key, value, required: false, max_length: 30, env_name: env_name)
+        when 'description', 'image', 'logo', 'repository', 'stack', 'success_url', 'website'
+          validate_string(key, value, required: false, env_name: env_name)
+        when 'environments'
+          validate_environments(key, value) if env_name.nil?
+        when 'keywords'
+          validate_array(key, value, required: false, env_name: env_name) do |v, index: nil, env_name: nil|
+            validate_string(key, v, required: false, index: index, env_name: env_name)
+          end
+        when 'scripts', 'formation'
+        else
+          raise InvalidManifest, "Unknown key: #{build_key_name(key, env_name: env_name)}"
```

Comment:
I think this could be dangerous to do. We may have customers that have old keys that were removed that would now fail. This would also limit us if in the future we wanted to remove or add a section that is optional/unknown today.